text/0000-dst-coercion.md

```diff
@@ -0,0 +1,182 @@
+- Feature Name: dst-coercions
+- Start Date: 2015-03-16
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Custom coercions allow smart pointers to fully participate in the DST system.
+In particular, they allow practical use of `Rc<T>` and `Arc<T>` where `T` is unsized.
+
+This RFC subsumes part of [RFC 401 coercions](https://github.com/rust-lang/rfcs/blob/master/text/0401-coercions.md).
+
+# Motivation
+
+DST is not really finished without this, in particular there is a need for types
+like reference counted trait objects (`Rc<Trait>`) which are not currently well-
+supported (without coercions, it is pretty much impossible to create such values
+with such a type).
+
+# Detailed design
+
+There is an `Unsize` trait and lang item. This trait signals that a type can be
+converted using the compiler's coercion machinery from a sized to an unsized
+type. All implementations of this trait are implicit and compiler generated. It
+is an error to implement this trait. If `&T` can be coerced to `&U` then there
+will be an implementation of `Unsize<U>` for `T`. E.g, `[i32; 42]:
+Unsize<[i32]>`. Note that the existence of an `Unsize` impl does not signify a
+coercion can itself can take place, it represents an internal part of the
+coercion mechanism (it corresponds with `coerce_inner` from  RFC 401). The trait
+is defined as:
```

Comment:
I don't think this design can prevent writing functions generic over unsizing coercions - is that even desired?